tests/topotests/ospf6_ecmp_inter_area/test_ospf6_ecmp_inter_area.py

```diff
@@ -0,0 +1,207 @@
+    for routern in range(1, 10):
+        tgen.add_router("r{}".format(routern))
+
+    tgen.gears["r1"].add_link(tgen.gears["r2"])
+    tgen.gears["r1"].add_link(tgen.gears["r3"])
+    tgen.gears["r1"].add_link(tgen.gears["r4"])
+    tgen.gears["r2"].add_link(tgen.gears["r5"])
+    tgen.gears["r3"].add_link(tgen.gears["r5"])
+    tgen.gears["r4"].add_link(tgen.gears["r5"])
+    tgen.gears["r5"].add_link(tgen.gears["r6"])
+    tgen.gears["r5"].add_link(tgen.gears["r7"])
+    tgen.gears["r5"].add_link(tgen.gears["r8"])
+    tgen.gears["r5"].add_link(tgen.gears["r9"])
+
+
+def setup_module(mod):
+    "Sets up the pytest environment"
+    tgen = Topogen(build_topo, mod.__name__)
+    tgen.start_topology()
+
+    router_list = tgen.routers()
+    for rname, router in router_list.items():
+        daemon_file = "{}/{}/zebra.conf".format(CWD, rname)
+        if os.path.isfile(daemon_file):
+            router.load_config(TopoRouter.RD_ZEBRA, daemon_file)
+
+        daemon_file = "{}/{}/ospf6d.conf".format(CWD, rname)
+        if os.path.isfile(daemon_file):
+            router.load_config(TopoRouter.RD_OSPF6, daemon_file)
+
+    # Initialize all routers.
+    tgen.start_router()
+
+
+def test_wait_protocol_convergence():
+    "Wait for OSPFv3 to converge"
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    logger.info("waiting for protocols to converge")
+
+    def expect_neighbor_full(router, neighbor):
+        "Wait until OSPFv3 convergence."
+        logger.info("waiting OSPFv3 router '{}'".format(router))
+        test_func = partial(
+            topotest.router_json_cmp,
+            tgen.gears[router],
+            "show ipv6 ospf6 neighbor json",
+            {"neighbors": [{"neighborId": neighbor, "state": "Full"}]},
+        )
+        _, result = topotest.run_and_expect(test_func, None, count=130, wait=1)
+        assertmsg = '"{}" convergence failure'.format(router)
+        assert result is None, assertmsg
+
+    expect_neighbor_full("r1", "10.254.254.2")
+    expect_neighbor_full("r1", "10.254.254.3")
+    expect_neighbor_full("r1", "10.254.254.4")
+    expect_neighbor_full("r2", "10.254.254.1")
+    expect_neighbor_full("r2", "10.254.254.5")
+    expect_neighbor_full("r3", "10.254.254.1")
+    expect_neighbor_full("r3", "10.254.254.5")
+    expect_neighbor_full("r4", "10.254.254.1")
+    expect_neighbor_full("r4", "10.254.254.5")
+    expect_neighbor_full("r5", "10.254.254.2")
+    expect_neighbor_full("r5", "10.254.254.3")
+    expect_neighbor_full("r5", "10.254.254.4")
+    expect_neighbor_full("r5", "10.254.254.6")
+    expect_neighbor_full("r5", "10.254.254.7")
+    expect_neighbor_full("r5", "10.254.254.8")
+    expect_neighbor_full("r5", "10.254.254.9")
+    expect_neighbor_full("r6", "10.254.254.5")
+    expect_neighbor_full("r7", "10.254.254.5")
+    expect_neighbor_full("r8", "10.254.254.5")
+    expect_neighbor_full("r9", "10.254.254.5")
+    topotest.sleep(10)
```

Comment:
What exactly is the problem with the sleep? Are you worried that the test is running too slow due to the sleeps? This one is relatively conservative and probably could be shortened, but I can't drop it completely because routes might not have been fully installed right after all neighborships are up. And the other one in the test needs to match the OSPF timers and can't be much shorter.
What do you suggest instead of the sleeps? Loops with repeated polling and checking for the expected results? I could do that, but it won't be much faster as the delay is dictated by the OSPF timers. And it will be less reliable because during convergence, you might have a short time where you get the expected result (and then terminate the loop) but if you'd wait a bit longer, you'd probably see that that's no longer the case after convergence.

Reply:
The problem with sleeps is that we have had this pattern employed by people writinig tests:
a) configure some change in state
b) sleep some small time
c) Assume state was correct and start lookinig for other end results

Not a big deal for a system that is not loaded or being shared with other tests.  But under high load we've seen a bunch of tests fail immediately after the sleep because the system did not actually converge to the new state.

Typically in the past people just made the sleep really large.  This caused really slow test run times.  Say we add a new peer for bgp.  A bgp timeout / retry for that peer cycle is typically 120 seconds.  So people were putting 120 second sleeps.  Now imagine that bgp actually converged in 5 seconds.  We just sat the test around for ~115 seconds.  In another attempt suppose the first packets failed to arrive ( due to system load/ whatever ).  A sleep of 120 seconds is insufficient and the test will fail.

What the community has settled on is, make some configuration change.  Then write a small loop of some sort that looks for that state change to have accepted/propagated.   In the bgp example above we typically add a new neighbor, then have a loop that checks for neighbor convergence and if it does not, we sleep for 1-2 seconds and then reissue the command.  This loop would time out after say 130 seconds.  This way the test is looking for convergence but not blindly hoping it happens.  And if we hit a timeout something has gone wrong and we know what has gone wrong.  As opposed to a system that blindly puts a sleep in and then looks for some end result ( say the addition of a route to the rib )

Reply:
Thanks for the explanation. Pushed an update without the sleeps.

OTOH, my previous version wasn't the simple "configure, sleep, check" version either. I'm first waiting for all neighborships to be as expected (that's probably the major source of delay and it won't go away, also in the new version). And then the old version had relatively small but still conservative sleeps to allow the network to reach post-convergence state before checking anything. So actually a combination of the two variants you described.

The new version now drops the sleeps and just waits for expected results. In theory, this could mean that we're not checking post-convergence state if the expected result occurs temporarily during convergence. In practice (at least with the bug this test case is checking), this shouldn't be an issue, though.